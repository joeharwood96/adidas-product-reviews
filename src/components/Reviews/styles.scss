.reviews {
    padding: 2rem;
    width: 60vw;
    margin: auto;

    @media (max-width:960px)  {
        width: auto;
    }

    &-title {
        text-transform: uppercase;
    }

    &-add {
        height: 2rem;
        border: none;
        border-bottom: 1px solid black;
        white-space: unset;
        text-transform: uppercase;
        background-color: white;
        display: flex;
        align-items: center;

        &:hover {
            cursor: pointer;
            background-color: black;
            color: white;
        }
    }

    &-average {
        background-color: rgb(42, 218, 113);
        padding: 2rem;
        text-align: center;
    }

    .on {
        color: black;
    }

    .off {
        color: gray;
    }
}