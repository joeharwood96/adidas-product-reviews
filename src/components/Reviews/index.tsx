import React, { useState } from 'react'
import './styles.scss'
import { useParams } from 'react-router-dom'
import { useReviewsQuery } from '../../services/productApi'
import { Review } from '../../models/review.model'

// Components
import Loader from '../Loader'
import AddReviewModal from '../AddReviewModal'
import Error from '../Error'

export default function Reviews() {
    const { id } = useParams();
    const [showAddReview, setShowAddReview] = useState(false);
    const { data, error, isLoading, isSuccess } = useReviewsQuery(id || '')
    const roundHalf = (num: number) => {
        return Math.round(num*2)/2;
    }

    return (
        <div className="reviews">
            <h2 className="reviews-title">Ratings and Reviews</h2>
            { isLoading && <Loader />}
            { error && <Error />}
            { isSuccess && data && 
                <div className="reviews-header">
                    <div style={{display: 'flex', alignItems: 'center', justifyContent: 'space-between'}}>
                    <h4>{data.length} {data.length === 1 ? "Review" : "Reviews"}</h4>
                    <button className="reviews-add" onClick={() => setShowAddReview(true)}>
                        <span className="reviews-add-text">Add Review</span>
                    </button>
                    </div>
                    {
                        data.length && <h4 className="reviews-average">Average {roundHalf(data.reduce((acc, cur) => cur.rating + acc, 0) / data.length)}</h4>
                    }
                </div>
            }

            { isSuccess && data?.map((review: Review, index) => 
                (
                    <div className="reviews-list" key={index}>
                        <h4>Review {index + 1}</h4>
                        {[...Array(5)].map((star, index) => {
                            return (<span key={index} className={`star ${index < review.rating ? "on" : "off"}`}>&#9733;</span>)
                        })}
                        <p>{review.text}</p>
                    </div>
                )
            )}

            {showAddReview && id && <AddReviewModal productId={id} setShowAddReview={setShowAddReview} />}
        </div>
    )
}