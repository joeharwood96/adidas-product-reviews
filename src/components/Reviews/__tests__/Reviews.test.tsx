import React from "react";
import Reviews from "../index";

describe("<Reviews />", () => {
    test("should display the Reviews", () => {
        const component = <Reviews />
        expect(component).toMatchSnapshot()
    });
});