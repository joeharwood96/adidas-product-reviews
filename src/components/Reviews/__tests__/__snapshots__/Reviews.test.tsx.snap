// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reviews /> should display the Reviews 1`] = `<Reviews />`;
